File: js-cplr/src/composer.ts
import { dirname, resolve } from 'node:path';
import type {
  ComponentFile,
  IncludeResolver,
  JsonObject,
  JsonValue,
  TokenResolver,
} from './types.js';
import { Validator } from './validator.js';

const REF_KEY = '.ref';
const VARS_KEY = '.vars';
const DEFAULTS_KEY = '.defaults';
const CONTENT_KEY = '.content';
const TOKEN_KEY = '.token';
const MISSING = Symbol('missing component variable');

const GROUP_NAME = /^[A-Za-z0-9_-]+$/;

// `@{group.path}` — the inline form of `.token`, usable anywhere a string is.
// `@@{` escapes a literal `@{`. Both are consumed at compile time, so the sigil
// is invisible to the client and composes inside runtime `${...}` expressions.
const TOKEN_SIGIL = /@@\{|@\{([^{}]*)\}/g;
const WHOLE_TOKEN_SIGIL = /^(?:@@\{|@\{([^{}]*)\})$/;

interface ExpansionContext {
  activePaths: readonly string[];
  componentStack: readonly string[];
  includeResolver: IncludeResolver;
  includeStack: readonly string[];
  referrerDir: string;
  rootDir: string;
  tokenResolver: TokenResolver;
}

type SubstitutionResult = JsonValue | typeof MISSING;

export class Composer {
  /**
   * Expands a parsed screen body into client-owned JSON.
   *
   * @param rootBody - Parsed screen template node tree
   * @param includeResolver - Injected path resolver and YAML loader
   * @param tokenResolver - Injected design-token group resolver
   * @param rootDir - SDUI root used to normalize resolved include paths
   * @param referrerDir - Directory containing the screen root file
   * @returns Fully expanded JSON without build-time keys
   * @throws When references, component interfaces, or reserved fields are invalid
   */
  static expand(
    rootBody: JsonValue,
    includeResolver: IncludeResolver,
    tokenResolver: TokenResolver,
    rootDir: string,
    referrerDir: string,
  ): JsonValue {
    return Composer.expandNode(rootBody, {
      activePaths: [],
      componentStack: [],
      includeResolver,
      includeStack: [],
      referrerDir,
      rootDir: resolve(rootDir),
      tokenResolver,
    });
  }

  /**
   * Substitutes exact component variable values and drops omitted slots.
   *
   * @param node - Component body node to transform
   * @param vars - Variable names declared by the component
   * @param args - Values supplied by the component reference
   * @returns Substituted JSON, or undefined when the root slot is omitted
   * @throws When the body contains an undeclared variable slot
   */
  static substituteVars(
    node: JsonValue,
    vars: readonly string[],
    args: Readonly<Record<string, JsonValue>>,
  ): JsonValue | undefined {
    const substituted = Composer.substituteNode(node, vars, args);
    return substituted === MISSING ? undefined : substituted;
  }

  /**
   * Substitutes array slots while splicing list-valued arguments in place.
   *
   * @param nodes - Component body array containing potential variable slots
   * @param vars - Variable names declared by the component
   * @param args - Values supplied by the component reference
   * @returns Substituted array with omitted slots removed
   * @throws When the body contains an undeclared variable slot
   */
  static spliceSlots(
    nodes: readonly JsonValue[],
    vars: readonly string[],
    args: Readonly<Record<string, JsonValue>>,
  ): JsonValue[] {
    const result: JsonValue[] = [];

    for (const node of nodes) {
      const slotName = Composer.slotName(node);
      if (slotName !== undefined) {
        Validator.assertHoleDeclared(vars, slotName);
        if (!Object.hasOwn(args, slotName)) continue;

        const value = Composer.clone(args[slotName]);
        if (Array.isArray(value)) result.push(...value);
        else result.push(value);
        continue;
      }

      const substituted = Composer.substituteNode(node, vars, args);
      if (substituted !== MISSING) result.push(substituted);
    }

    return result;
  }

  private static expandNode(
    node: JsonValue,
    context: ExpansionContext,
  ): JsonValue {
    if (Array.isArray(node)) return Composer.expandArray(node, context);
    if (typeof node === 'string') return Composer.expandString(node, context);
    if (!Composer.isObject(node)) return node;

    if (Object.hasOwn(node, 'screen_id')) {
      throw new Error('screen_id is not allowed in template nodes');
    }
    if (Object.hasOwn(node, TOKEN_KEY)) {
      return Composer.expandToken(node, context);
    }
    if (Object.hasOwn(node, REF_KEY)) {
      return Composer.expandReference(node, context);
    }

    const result: JsonObject = {};
    for (const [key, value] of Object.entries(node)) {
      if (key.startsWith('.')) {
        throw new Error(`Unsupported build key ${key}`);
      }
      result[key] = Composer.expandNode(value, context);
    }
    return result;
  }

  private static expandToken(
    node: JsonObject,
    context: ExpansionContext,
  ): JsonValue {
    if (Object.keys(node).length !== 1) {
      throw new Error('.token node must not have sibling keys');
    }

    const tokenPath = node[TOKEN_KEY];
    if (typeof tokenPath !== 'string') {
      throw new Error('.token value must be a non-empty dotted string');
    }

    return Composer.resolveToken(tokenPath, context);
  }

  /**
   * Resolves a dotted token path against its group file.
   *
   * Shared by `{ .token: ... }` and the inline `@{...}` sigil so both forms
   * validate identically and fail with the same diagnostics.
   *
   * @param tokenPath - Dotted path, `<group>.<key...>`
   * @param context - Expansion context carrying the token resolver
   * @returns Deep copy of the token value
   * @throws When the path is malformed, the group is unknown, or a key is missing
   */
  private static resolveToken(
    tokenPath: string,
    context: ExpansionContext,
  ): JsonValue {
    const segments = tokenPath.split('.');
    if (
      segments.length < 2 ||
      segments.some(
        (segment) => segment.length === 0 || segment.trim() !== segment,
      ) ||
      !GROUP_NAME.test(segments[0])
    ) {
      throw new Error('.token value must be a non-empty dotted string');
    }

    const [group, ...keys] = segments;
    let value: JsonValue = context.tokenResolver(group);
    let traversedPath = group;

    for (const key of keys) {
      if (!Composer.isObject(value)) {
        throw new Error(
          `Cannot traverse non-object token path ${traversedPath} while resolving ${tokenPath}`,
        );
      }
      if (!Object.hasOwn(value, key)) {
        throw new Error(
          `Missing token key ${tokenPath} at ${traversedPath}.${key}`,
        );
      }

      value = value[key];
      traversedPath = `${traversedPath}.${key}`;
    }

    // A token value that carries the sigil would be rescanned when it lands in
    // a component argument subtree, making resolution depend on where it was
    // used. Reject it at the source instead.
    if (typeof value === 'string' && value.includes('@{')) {
      throw new Error(`Token value must not contain @{: ${tokenPath}`);
    }

    return Composer.clone(value);
  }

  /**
   * Substitutes `@{group.path}` token sigils inside a string value.
   *
   * A string that is exactly one sigil resolves to the token's own value and
   * keeps its type (`'@{spacing.lg}'` -> `16`). Any other occurrence is
   * interpolated as text, so tokens compose inside runtime expressions.
   *
   * @param node - Raw string value from the template
   * @param context - Expansion context carrying the token resolver
   * @returns The token value, the interpolated string, or the string unchanged
   * @throws When a sigil is unterminated or interpolates a non-scalar token
   */
  private static expandString(
    node: string,
    context: ExpansionContext,
  ): JsonValue {
    if (!node.includes('@')) return node;

    const whole = WHOLE_TOKEN_SIGIL.exec(node);
    if (whole !== null && whole[1] !== undefined) {
      return Composer.resolveToken(whole[1], context);
    }

    Composer.assertNoDanglingSigil(node);
    return node.replace(TOKEN_SIGIL, (_match, path?: string) => {
      if (path === undefined) return '@{';

      const value = Composer.resolveToken(path, context);
      if (Composer.isObject(value) || Array.isArray(value)) {
        throw new Error(
          `Cannot interpolate non-scalar token ${path} into a string`,
        );
      }
      return typeof value === 'string' ? value : JSON.stringify(value);
    });
  }

  /**
   * Rejects an unterminated `@{` left behind by substitution.
   *
   * A typo like `@{color.primary` would otherwise ship to the client as
   * literal text; failing the build is the loud alternative.
   */
  private static assertNoDanglingSigil(original: string): void {
    if (original.replace(TOKEN_SIGIL, '').includes('@{')) {
      throw new Error(`Unterminated token sigil @{ in: ${original}`);
    }
  }

  private static expandArray(
    nodes: readonly JsonValue[],
    context: ExpansionContext,
  ): JsonValue[] {
    const result: JsonValue[] = [];

    for (const node of nodes) {
      const expanded = Composer.expandNode(node, context);
      if (Composer.isReference(node) && Array.isArray(expanded)) {
        result.push(...expanded);
      } else {
        result.push(expanded);
      }
    }

    return result;
  }

  private static expandReference(
    node: JsonObject,
    context: ExpansionContext,
  ): JsonValue {
    if (Object.hasOwn(node, '_type')) {
      throw new Error('.ref node must not declare _type');
    }

    const refPath = node[REF_KEY];
    if (typeof refPath !== 'string') {
      throw new Error('.ref path must be a string');
    }

    const resolvedInclude = context.includeResolver(
      refPath,
      context.referrerDir,
    );
    const filePath = resolve(context.rootDir, resolvedInclude.filePath);
    Validator.assertNoCircularPath(filePath, context.activePaths);

    // Component args are authored by the caller, so expand them in the caller's
    // context — before this file is pushed onto activePaths. Otherwise reusing a
    // component inside another instance's arg subtree (e.g. a /card in a /card's
    // child) re-enters the same file while it is still "active" and is misflagged
    // as circular. Genuine self-recursion is still caught: a component whose own
    // .content re-references itself expands that .content with the file on the path.
    const args = Object.fromEntries(
      Object.entries(node)
        .filter(([key]) => key !== REF_KEY)
        .map(([key, value]) => [key, Composer.expandNode(value, context)]),
    );
    const targetContext = {
      ...context,
      activePaths: [...context.activePaths, filePath],
      referrerDir: dirname(filePath),
    };

    if (Composer.isComponent(resolvedInclude.content)) {
      return Composer.expandComponent(
        resolvedInclude.content,
        args,
        filePath,
        targetContext,
      );
    }

    const argNames = Object.keys(args);
    if (argNames.length > 0) {
      throw new Error(
        `Fragment ${filePath} does not accept args: ${argNames.join(', ')}`,
      );
    }

    return Composer.expandNode(resolvedInclude.content, {
      ...targetContext,
      includeStack: [...context.includeStack, filePath],
    });
  }

  private static expandComponent(
    target: JsonObject,
    args: Readonly<Record<string, JsonValue>>,
    filePath: string,
    context: ExpansionContext,
  ): JsonValue {
    const component = Composer.toComponentFile(target, filePath);
    Validator.assertArgsDeclared(component.vars, args);
    const effectiveArgs = { ...(component.defaults ?? {}), ...args };

    const substituted = Composer.substituteVars(
      component.content,
      component.vars,
      effectiveArgs,
    );
    if (substituted === undefined) {
      throw new Error(
        `Component body resolved to an omitted slot: ${filePath}`,
      );
    }

    return Composer.expandNode(substituted, {
      ...context,
      componentStack: [...context.componentStack, filePath],
    });
  }

  private static substituteNode(
    node: JsonValue,
    vars: readonly string[],
    args: Readonly<Record<string, JsonValue>>,
  ): SubstitutionResult {
    if (Array.isArray(node)) return Composer.spliceSlots(node, vars, args);

    if (Composer.isObject(node)) {
      const result: JsonObject = {};
      for (const [key, value] of Object.entries(node)) {
        if (key.startsWith('.') && key !== REF_KEY && key !== TOKEN_KEY) {
          throw new Error(`Unsupported build key ${key}`);
        }
        const substituted = Composer.substituteNode(value, vars, args);
        if (substituted !== MISSING) result[key] = substituted;
      }
      return result;
    }

    const slotName = Composer.slotName(node);
    if (slotName === undefined) return node;

    Validator.assertHoleDeclared(vars, slotName);
    if (!Object.hasOwn(args, slotName)) return MISSING;
    return Composer.clone(args[slotName]);
  }

  private static toComponentFile(
    target: JsonObject,
    filePath: string,
  ): ComponentFile {
    if (Object.hasOwn(target, 'screen_id')) {
      throw new Error('screen_id is not allowed in template nodes');
    }
    const unsupportedKey = Object.keys(target).find(
      (key) =>
        key.startsWith('.') &&
        key !== VARS_KEY &&
        key !== DEFAULTS_KEY &&
        key !== CONTENT_KEY,
    );
    if (unsupportedKey !== undefined) {
      throw new Error(`Unsupported build key ${unsupportedKey}`);
    }

    const vars = target[VARS_KEY];
    if (
      !Composer.isStringArray(vars) ||
      vars.some((name) => name.length === 0)
    ) {
      throw new Error(`Component .vars must be a list of names: ${filePath}`);
    }
    if (new Set(vars).size !== vars.length) {
      throw new Error(`Component .vars contains duplicate names: ${filePath}`);
    }
    const defaults = target[DEFAULTS_KEY];
    if (defaults !== undefined && !Composer.isObject(defaults)) {
      throw new Error(`Component .defaults must be a map: ${filePath}`);
    }
    if (defaults !== undefined) {
      for (const key of Object.keys(defaults)) {
        if (!vars.includes(key)) {
          throw new Error(
            `Component .defaults key not in .vars: ${key} (${filePath})`,
          );
        }
      }
    }
    if (!Object.hasOwn(target, CONTENT_KEY)) {
      throw new Error(
        `Component with .vars must declare .content: ${filePath}`,
      );
    }

    return { vars, defaults, content: target[CONTENT_KEY] };
  }

  private static slotName(node: JsonValue): string | undefined {
    if (
      typeof node !== 'string' ||
      !node.startsWith('.') ||
      node.length === 1
    ) {
      return undefined;
    }
    if (node.startsWith('./') || node.startsWith('..')) return undefined;
    return node.slice(1);
  }

  private static isReference(node: JsonValue): node is JsonObject {
    return Composer.isObject(node) && Object.hasOwn(node, REF_KEY);
  }

  private static isComponent(node: JsonValue): node is JsonObject {
    return Composer.isObject(node) && Object.hasOwn(node, VARS_KEY);
  }

  private static isStringArray(node: JsonValue): node is string[] {
    return (
      Array.isArray(node) && node.every((item) => typeof item === 'string')
    );
  }

  private static isObject(node: JsonValue): node is JsonObject {
    return typeof node === 'object' && node !== null && !Array.isArray(node);
  }

  private static clone(node: JsonValue): JsonValue {
    if (Array.isArray(node)) return node.map((item) => Composer.clone(item));
    if (!Composer.isObject(node)) return node;

    return Object.fromEntries(
      Object.entries(node).map(([key, value]) => [key, Composer.clone(value)]),
    );
  }
}

File: py-cplr/src/sdui_template_compiler/composer.py
"""Template expansion: references, components, tokens, and variable slots."""

import copy
import os
import re
from dataclasses import dataclass, replace
from typing import Any, Mapping, Sequence

from .include_resolver import IncludeResolverFn
from .json_compat import stringify
from .token_resolver import TokenResolverFn
from .validator import Validator

JsonValue = Any
JsonObject = dict[str, Any]

REF_KEY = ".ref"
VARS_KEY = ".vars"
DEFAULTS_KEY = ".defaults"
CONTENT_KEY = ".content"
TOKEN_KEY = ".token"

_MISSING = object()
_GROUP_NAME = re.compile(r"^[A-Za-z0-9_-]+$")

# `@{group.path}` — the inline form of `.token`, usable anywhere a string is.
# `@@{` escapes a literal `@{`. Both are consumed at compile time, so the sigil
# is invisible to the client and composes inside runtime `${...}` expressions.
_TOKEN_SIGIL = re.compile(r"@@\{|@\{([^{}]*)\}")


@dataclass(frozen=True)
class ComponentFile:
    vars: Sequence[str]
    defaults: Mapping[str, JsonValue] | None
    content: JsonValue


@dataclass(frozen=True)
class _ExpansionContext:
    active_paths: tuple[str, ...]
    component_stack: tuple[str, ...]
    include_resolver: IncludeResolverFn
    include_stack: tuple[str, ...]
    referrer_dir: str
    root_dir: str
    token_resolver: TokenResolverFn


class Composer:
    """Expands parsed screen templates into client-owned JSON."""

    @staticmethod
    def expand(
        root_body: JsonValue,
        include_resolver: IncludeResolverFn,
        token_resolver: TokenResolverFn,
        root_dir: str,
        referrer_dir: str,
    ) -> JsonValue:
        """Expands a parsed screen body into client-owned JSON.

        @param root_body - Parsed screen template node tree
        @param include_resolver - Injected path resolver and YAML loader
        @param token_resolver - Injected design-token group resolver
        @param root_dir - SDUI root used to normalize resolved include paths
        @param referrer_dir - Directory containing the screen root file
        @returns Fully expanded JSON without build-time keys
        @throws When references, component interfaces, or reserved fields are invalid
        """
        return Composer._expand_node(
            root_body,
            _ExpansionContext(
                active_paths=(),
                component_stack=(),
                include_resolver=include_resolver,
                include_stack=(),
                referrer_dir=referrer_dir,
                root_dir=os.path.abspath(root_dir),
                token_resolver=token_resolver,
            ),
        )

    @staticmethod
    def substitute_vars(
        node: JsonValue,
        vars: Sequence[str],
        args: Mapping[str, JsonValue],
    ) -> JsonValue | None:
        """Substitutes exact component variable values and drops omitted slots.

        @param node - Component body node to transform
        @param vars - Variable names declared by the component
        @param args - Values supplied by the component reference
        @returns Substituted JSON, or None when the root slot is omitted
        @throws When the body contains an undeclared variable slot
        """
        substituted = Composer._substitute_node(node, vars, args)
        return None if substituted is _MISSING else substituted

    @staticmethod
    def splice_slots(
        nodes: Sequence[JsonValue],
        vars: Sequence[str],
        args: Mapping[str, JsonValue],
    ) -> list[JsonValue]:
        """Substitutes array slots while splicing list-valued arguments in place.

        @param nodes - Component body array containing potential variable slots
        @param vars - Variable names declared by the component
        @param args - Values supplied by the component reference
        @returns Substituted array with omitted slots removed
        @throws When the body contains an undeclared variable slot
        """
        result: list[JsonValue] = []

        for node in nodes:
            slot_name = Composer._slot_name(node)
            if slot_name is not None:
                Validator.assert_hole_declared(vars, slot_name)
                if slot_name not in args:
                    continue

                value = copy.deepcopy(args[slot_name])
                if isinstance(value, list):
                    result.extend(value)
                else:
                    result.append(value)
                continue

            substituted = Composer._substitute_node(node, vars, args)
            if substituted is not _MISSING:
                result.append(substituted)

        return result

    @staticmethod
    def _expand_node(node: JsonValue, context: _ExpansionContext) -> JsonValue:
        if isinstance(node, list):
            return Composer._expand_array(node, context)
        if isinstance(node, str):
            return Composer._expand_string(node, context)
        if not isinstance(node, dict):
            return node

        if "screen_id" in node:
            raise ValueError("screen_id is not allowed in template nodes")
        if TOKEN_KEY in node:
            return Composer._expand_token(node, context)
        if REF_KEY in node:
            return Composer._expand_reference(node, context)

        result: JsonObject = {}
        for key, value in node.items():
            if key.startswith("."):
                raise ValueError(f"Unsupported build key {key}")
            result[key] = Composer._expand_node(value, context)
        return result

    @staticmethod
    def _expand_token(node: JsonObject, context: _ExpansionContext) -> JsonValue:
        if len(node) != 1:
            raise ValueError(".token node must not have sibling keys")

        token_path = node[TOKEN_KEY]
        if not isinstance(token_path, str):
            raise ValueError(".token value must be a non-empty dotted string")

        return Composer._resolve_token(token_path, context)

    @staticmethod
    def _resolve_token(token_path: str, context: _ExpansionContext) -> JsonValue:
        """Resolves a dotted token path against its group file.

        Shared by `{ .token: ... }` and the inline `@{...}` sigil so both forms
        validate identically and fail with the same diagnostics.

        @param token_path - Dotted path, `<group>.<key...>`
        @param context - Expansion context carrying the token resolver
        @returns Deep copy of the token value
        @throws When the path is malformed, the group is unknown, or a key is missing
        """
        segments = token_path.split(".")
        if (
            len(segments) < 2
            or any(len(segment) == 0 or segment.strip() != segment for segment in segments)
            or _GROUP_NAME.match(segments[0]) is None
        ):
            raise ValueError(".token value must be a non-empty dotted string")

        group, *keys = segments
        value: JsonValue = context.token_resolver(group)
        traversed_path = group

        for key in keys:
            if not isinstance(value, dict):
                raise ValueError(
                    f"Cannot traverse non-object token path {traversed_path} "
                    f"while resolving {token_path}"
                )
            if key not in value:
                raise ValueError(f"Missing token key {token_path} at {traversed_path}.{key}")

            value = value[key]
            traversed_path = f"{traversed_path}.{key}"

        # A token value that carries the sigil would be rescanned when it lands in
        # a component argument subtree, making resolution depend on where it was
        # used. Reject it at the source instead.
        if isinstance(value, str) and "@{" in value:
            raise ValueError(f"Token value must not contain @{{: {token_path}")

        return copy.deepcopy(value)

    @staticmethod
    def _expand_string(node: str, context: _ExpansionContext) -> JsonValue:
        """Substitutes `@{group.path}` token sigils inside a string value.

        A string that is exactly one sigil resolves to the token's own value and
        keeps its type (`'@{spacing.lg}'` -> `16`). Any other occurrence is
        interpolated as text, so tokens compose inside runtime expressions.

        @param node - Raw string value from the template
        @param context - Expansion context carrying the token resolver
        @returns The token value, the interpolated string, or the string unchanged
        @throws When a sigil is unterminated or interpolates a non-scalar token
        """
        if "@" not in node:
            return node

        whole = _TOKEN_SIGIL.fullmatch(node)
        if whole is not None and whole.group(1) is not None:
            return Composer._resolve_token(whole.group(1), context)

        def substitute(match: re.Match[str]) -> str:
            if match.group(1) is None:
                return "@{"

            value = Composer._resolve_token(match.group(1), context)
            if isinstance(value, (dict, list)):
                raise ValueError(
                    f"Cannot interpolate non-scalar token {match.group(1)} into a string"
                )
            return value if isinstance(value, str) else stringify(value)

        Composer._assert_no_dangling_sigil(node)
        return _TOKEN_SIGIL.sub(substitute, node)

    @staticmethod
    def _assert_no_dangling_sigil(original: str) -> None:
        """Rejects an unterminated `@{` left behind by substitution.

        A typo like `@{color.primary` would otherwise ship to the client as
        literal text; failing the build is the loud alternative.
        """
        if "@{" in _TOKEN_SIGIL.sub("", original):
            raise ValueError(f"Unterminated token sigil @{{ in: {original}")

    @staticmethod
    def _expand_array(nodes: Sequence[JsonValue], context: _ExpansionContext) -> list[JsonValue]:
        result: list[JsonValue] = []

        for node in nodes:
            expanded = Composer._expand_node(node, context)
            if Composer._is_reference(node) and isinstance(expanded, list):
                result.extend(expanded)
            else:
                result.append(expanded)

        return result

    @staticmethod
    def _expand_reference(node: JsonObject, context: _ExpansionContext) -> JsonValue:
        if "_type" in node:
            raise ValueError(".ref node must not declare _type")

        ref_path = node[REF_KEY]
        if not isinstance(ref_path, str):
            raise ValueError(".ref path must be a string")

        resolved_include = context.include_resolver(ref_path, context.referrer_dir)
        file_path = os.path.abspath(os.path.join(context.root_dir, resolved_include.file_path))
        Validator.assert_no_circular_path(file_path, context.active_paths)

        # Component args are authored by the caller, so expand them in the caller's
        # context — before this file is pushed onto active_paths. Otherwise reusing a
        # component inside another instance's arg subtree (e.g. a /card in a /card's
        # child) re-enters the same file while it is still "active" and is misflagged
        # as circular. Genuine self-recursion is still caught: a component whose own
        # .content re-references itself expands that .content with the file on the path.
        args = {
            key: Composer._expand_node(value, context)
            for key, value in node.items()
            if key != REF_KEY
        }
        target_context = replace(
            context,
            active_paths=(*context.active_paths, file_path),
            referrer_dir=os.path.dirname(file_path),
        )

        if Composer._is_component(resolved_include.content):
            return Composer._expand_component(
                resolved_include.content, args, file_path, target_context
            )

        if args:
            raise ValueError(f"Fragment {file_path} does not accept args: {', '.join(args)}")

        return Composer._expand_node(
            resolved_include.content,
            replace(target_context, include_stack=(*context.include_stack, file_path)),
        )

    @staticmethod
    def _expand_component(
        target: JsonObject,
        args: Mapping[str, JsonValue],
        file_path: str,
        context: _ExpansionContext,
    ) -> JsonValue:
        component = Composer._to_component_file(target, file_path)
        Validator.assert_args_declared(component.vars, args)
        effective_args = {**(component.defaults or {}), **args}

        substituted = Composer.substitute_vars(component.content, component.vars, effective_args)
        if substituted is None:
            raise ValueError(f"Component body resolved to an omitted slot: {file_path}")

        return Composer._expand_node(
            substituted,
            replace(context, component_stack=(*context.component_stack, file_path)),
        )

    @staticmethod
    def _substitute_node(
        node: JsonValue,
        vars: Sequence[str],
        args: Mapping[str, JsonValue],
    ) -> JsonValue:
        if isinstance(node, list):
            return Composer.splice_slots(node, vars, args)

        if isinstance(node, dict):
            result: JsonObject = {}
            for key, value in node.items():
                if key.startswith(".") and key != REF_KEY and key != TOKEN_KEY:
                    raise ValueError(f"Unsupported build key {key}")
                substituted = Composer._substitute_node(value, vars, args)
                if substituted is not _MISSING:
                    result[key] = substituted
            return result

        slot_name = Composer._slot_name(node)
        if slot_name is None:
            return node

        Validator.assert_hole_declared(vars, slot_name)
        if slot_name not in args:
            return _MISSING
        return copy.deepcopy(args[slot_name])

    @staticmethod
    def _to_component_file(target: JsonObject, file_path: str) -> ComponentFile:
        if "screen_id" in target:
            raise ValueError("screen_id is not allowed in template nodes")
        unsupported_key = next(
            (
                key
                for key in target
                if key.startswith(".")
                and key not in (VARS_KEY, DEFAULTS_KEY, CONTENT_KEY)
            ),
            None,
        )
        if unsupported_key is not None:
            raise ValueError(f"Unsupported build key {unsupported_key}")

        vars = target[VARS_KEY]
        if not Composer._is_string_list(vars) or any(len(name) == 0 for name in vars):
            raise ValueError(f"Component .vars must be a list of names: {file_path}")
        if len(set(vars)) != len(vars):
            raise ValueError(f"Component .vars contains duplicate names: {file_path}")
        defaults = target[DEFAULTS_KEY] if DEFAULTS_KEY in target else None
        if DEFAULTS_KEY in target and not isinstance(defaults, dict):
            raise ValueError(f"Component .defaults must be a map: {file_path}")
        if defaults is not None:
            for key in defaults:
                if key not in vars:
                    raise ValueError(f"Component .defaults key not in .vars: {key} ({file_path})")
        if CONTENT_KEY not in target:
            raise ValueError(f"Component with .vars must declare .content: {file_path}")

        return ComponentFile(vars=vars, defaults=defaults, content=target[CONTENT_KEY])

    @staticmethod
    def _slot_name(node: JsonValue) -> str | None:
        if not isinstance(node, str) or not node.startswith(".") or len(node) == 1:
            return None
        if node.startswith("./") or node.startswith(".."):
            return None
        return node[1:]

    @staticmethod
    def _is_reference(node: JsonValue) -> bool:
        return isinstance(node, dict) and REF_KEY in node

    @staticmethod
    def _is_component(node: JsonValue) -> bool:
        return isinstance(node, dict) and VARS_KEY in node

    @staticmethod
    def _is_string_list(node: JsonValue) -> bool:
        return isinstance(node, list) and all(isinstance(item, str) for item in node)
